feat: update the docs for edge function testing (#47055)
supabase/supabase: apps/docs/content/guides/functions/unit-test.mdx

subtitle: 'Writing Unit Tests for Edge Functions using Deno Test'
---

> # Testing your Edge Functions

Reviewer: Since `title` adds the `h1`, I recommend making this an `h2` and so on.

Reviewer: Additionally, this duplicates the title. You can name this "Overview" instead.

Reviewer 2: solved at ac6d5f4c5a61

Testing is an essential step in the development process to ensure the correctness, reliability, and performance of your Edge Functions. Because Edge Functions often combine HTTP handling, authentication, database access, and business logic, a good testing strategy gives you fast feedback and high confidence before deploying to production.

In this guide you will learn how to write:

> - **Unit tests** for pure business logic (pricing rules, calculations, etc.)

Reviewer: Parenthetical is typically reserved for defining acronyms. 
```suggestion
- **Unit tests** for pure business logic such as pricing rules, calculations, etc.
```

Reviewer 2: Solved at db7251fc772b

- **Integration tests** for the full Edge Function by mocking at the network layer

> The examples and patterns shown here come directly from the reference implementation in the [play.supabase.unit-tests](https://github.com/kallebysantos/play.supabase.unit-tests) repository and follow the same approaches used internally by Supabase's Edge Functions team.

Reviewer: A bit verbose. Can be simplified.

```suggestion
You can also find the examples in this example repository: [play.supabase.unit-tests](https://github.com/kallebysantos/play.supabase.unit-tests)
```

To "by Supabase's Edge Functions team," it should be implied to be official when this merges in the Supabase org, hopefully.

Reviewer 2: Solved at 136b85cff0c8

Deno ships with a fast, native test runner and excellent mocking utilities in `@std/testing`. See the [official Deno testing documentation](https://docs.deno.com/runtime/manual/basics/testing/) for more background.

---

## The example scenario

> We will use a realistic Edge Function called `process-ticket` that calculates the final price of a ticket based on the authenticated user's age (loaded from the `profiles` table).

Reviewer: - Use present tense
- Use 2nd person singular

For example:
```suggestion
You can use a realistic Edge Function called `process-ticket` that calculates the final price of a ticket based on the authenticated user's age (loaded from the `profiles` table).
```

Reviewer 2: solved at 7cde06c790d0

**Business rules:**

- Children aged 8 and under → free (`0`)
- Young people aged 9–17 → 20% discount
- Adults aged 18 and over → full price

The function receives a JSON payload with a `price` field and returns `{ result: finalPrice }`.

This example demonstrates common real-world requirements:

- Request validation
- Authenticated database access via `withSupabase`
- Business rule application
- Proper error handling

---

## Recommended project structure

```
supabase/
├── functions/
│   ├── _shared/
│   │   └── types.ts                 # Database types
│   ├── process-ticket/
│   │   ├── index.ts                 # Edge Function (uses withSupabase)
│   │   └── pricing.ts               # Pure business logic (co-located)
│   └── tests/
│       ├── utils/
│       │   └── supabase_env.ts      # Test helpers (env + JWT)
│       └── process-ticket/
│           ├── pricing.test.ts      # Unit tests for pricing
│           └── index.test.ts        # Integration tests with fetch mocking
├── config.toml
└── deno.json
```

> > **Note:** In this reference implementation the pricing logic lives inside the function folder (`process-ticket/pricing.ts`). You can also move it to `_shared/` if you want to reuse it across multiple functions.

Reviewer: I think we have a convention for this.

!!! note

See the [Development Environment](/docs/guides/functions/development-environment) and [Managing dependencies](/docs/guides/functions/dependencies) guides for recommended `deno.json` and editor setup.

---

## Unit tests: Testing pure business logic

The pricing rules are pure functions with no side effects, so they are perfect candidates for fast, isolated unit tests.

### The pricing module (`process-ticket/pricing.ts`)

```ts
// supabase/functions/process-ticket/pricing.ts
export const AGE_TIERS = {
  CHILDREN: 1, // free access
  YOUNG: 0.2, // 20% off (ages 9-17)
  ADULT: 0, // no discount
}

export function getAgeDiscount(age: number) {
  if (age <= 8) {
    return AGE_TIERS.CHILDREN
  }

  if (age > 8 && age < 18) {
    return AGE_TIERS.YOUNG
  }

  return AGE_TIERS.ADULT
}

export function applyTicketDiscount(price: number, age: number) {
  const discount = getAgeDiscount(age)
  return price - price * discount
}
```

> ### Unit tests (`tests/process-ticket/pricing.test.ts`)

Reviewer: I'm unsure about this parenthetical here. Headers typically direct the eye but does not give information directly.

Reviewer 2: Solved at 1ba749e9a59b

        Authorization: `Bearer ${token}`,
      },
      body: JSON.stringify({ price: 15 }),
    })

    const mockProfiles = mockTableRequest('profiles?select=age&limit=1', {})
    globalThis.fetch = mockProfiles

    const res = await EdgeFunction.fetch(req)
    assertEquals(res.status, 500)
    assertEquals(await res.json(), { error: 'could not process' })
  })

  it('should handle database errors gracefully', async () => {
    const token = await generateUserToken()
    const req = new Request(FUNCTION_URL, {
      method: 'POST',
      headers: {
        apikey: env.publishableKeys.default,
        Authorization: `Bearer ${token}`,
      },
      body: JSON.stringify({ price: 15 }),
    })

    const mockProfiles = mockTableRequest(
      'profiles?select=age&limit=1',
      { code: 'PGRST303', message: 'JWT expired' },
      401
    )
    globalThis.fetch = mockProfiles

    const res = await EdgeFunction.fetch(req)
    assertEquals(res.status, 500)
    assertEquals(await res.json(), { error: 'could not process' })
  })
})
```

Run the integration tests:

```bash
deno test supabase/functions/tests/process-ticket/index.test.ts --allow-env
```

---

> ## Why this mocking approach is by mocking `globalThis.fetch`, you:

Reviewer: Keep headings short. Should probably not include an `:`

Reviewer 2: Solved at 8a22d2276c39

> - Test the **real** Edge Function code path (no dependency injection needed in production code)

Reviewer: Again, avoid parenthetical

- Simulate database responses, auth failures, network errors, etc.
- Keep your production Edge Function clean and focused
- Still get fast, deterministic tests that don't require a running Supabase instance

> This pattern is especially useful when using higher-level helpers like `withSupabase` that create the Supabase client internally.

Reviewer: Avoid marketing language such as "especially useful". Just say what it does.

---

## Running all tests

Add to your `deno.json`:

```json
{
  "tasks": {
    "test": "deno test supabase/functions/tests/ --allow-env"
  }
}
```

Then:

```bash
deno task test
```

---

> ## Best practices demonstrated in this guide

Reviewer: Keep Short. Maybe just "Best practices"

Reviewer 2: solved at ea9988d5fea4
